python/ray/serve/pipeline/json_serde.py

```diff
@@ -0,0 +1,166 @@
+from typing import Any, Union
+from importlib import import_module
+
+import json
+
+from ray.experimental.dag import (
+    DAGNode,
+    ClassNode,
+    ClassMethodNode,
+    FunctionNode,
+    InputNode,
+)
+from ray.serve.api import Deployment, DeploymentConfig
+from ray.serve.pipeline.deployment_node import DeploymentNode
+from ray.serve.pipeline.deployment_method_node import DeploymentMethodNode
+from ray.serve.utils import parse_import_path
+from ray.serve.handle import RayServeHandle
+from ray.serve.utils import ServeHandleEncoder, serve_handle_object_hook
+from ray.serve.constants import (
+    SERVE_ASYNC_HANDLE_JSON_PREFIX_KEY,
+    SERVE_SYNC_HANDLE_JSON_PREFIX_KEY,
+)
+
+# Reserved key to distinguish DAGNode type and avoid collision with user dict.
+DAGNODE_TYPE_KEY = "__serve_pipeline_dag_node_type__"
+
+
+class DAGNodeEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, RayServeHandle):
+            return json.dumps(obj, cls=ServeHandleEncoder)
+        if isinstance(obj, DAGNode):
+            dag_node = obj
+            args = dag_node.get_args()
+            kwargs = dag_node.get_kwargs()
+            options = dag_node.get_options()
+            other_args_to_resolve = dag_node.get_other_args_to_resolve()
+            # stable_uuid will be re-generated upon new constructor execution
+            result_dict = {
+                DAGNODE_TYPE_KEY: dag_node.__class__.__name__,
+                "args": json.dumps(args, cls=DAGNodeEncoder),
+                "kwargs": json.dumps(kwargs, cls=DAGNodeEncoder),
+                # .options() should not contain any DAGNode type
+                "options": json.dumps(options),
+                "other_args_to_resolve": json.dumps(
+                    other_args_to_resolve, cls=DAGNodeEncoder
+                ),
+            }
+            # TODO: (jiaodong) Support arbitrary InputNode args and pydantic
+            # input schema.
+            import_path = ""
+            if isinstance(dag_node, InputNode):
+                return result_dict
+            # TODO: (jiaodong) Handle __main__ swap to filename
+            if isinstance(dag_node, ClassNode):
+                body = dag_node._body.__ray_actor_class__
+                import_path = f"{body.__module__}.{body.__qualname__}"
+            elif isinstance(dag_node, FunctionNode):
+                body = dag_node._body
+                import_path = f"{body.__module__}.{body.__qualname__}"
+            elif isinstance(dag_node, (DeploymentNode, DeploymentMethodNode)):
+                result_dict.update(
+                    {"deployment_name": dag_node._deployment.name}
+                )
+                if isinstance(dag_node._deployment._func_or_class, str):
+                    # We're processing a deserilized JSON node where import_path
+                    # is dag_node body.
```

Comment:
please make the `import_path` logic a standalone utility and add unit tests for it.

Reply:
i moved these to each DAGNode type now, will address in later diff before beta when we polish generated yaml part. We can also cover with some remote runtime_env tests to properly test all import_path related items.